kubespawner/utils.py

```diff
@@ -45,3 +48,82 @@ def validate(self, obj, value):
             return value
         else:
             self.error(obj, value)
+
+@gen.coroutine
+def ensure_object(
+    asynchronize_func,
+    create_func,
+    patch_func,
+    body,
+    ifexists='patch',
+    namespace=None
+):
+    """
+    Ensure a given k8s object exists
+
+    create_func - the API function for creating the object
+    patch_fun - API method for patching the object (if required)
+    body - the object to be created
+    ifexists - what to do if the object already exists.
+               Supported options are 'patch' and 'ignore'
+    namespace - Namespace to create object in, if it is a namespaced object
+    """
+    name = body.metadata.name
+    try:
+        create_kwargs = {'body': body}
+        if namespace:
+            create_kwargs['namespace'] = namespace
+        resp = yield asynchronize_func(
+            create_func,
+            **create_kwargs
+        )
+    except client.rest.ApiException as e:
+        if e.status == 409:
+            if ifexists == 'patch':
+                # This object already exists, we should patch it to make it be what we want
+                patch_kwargs = {'body': body, 'name': body.metadata.name}
+                if namespace:
+                    patch_kwargs['namespace'] = namespace
+                resp = yield asynchronize_func(
+                    patch_func,
+                    **patch_kwargs
+                )
+            elif ifexists == 'ignore':
+                pass
+            else:
+                raise
+        else:
+            raise
+def first_upper(s):
+    """
+    Upper case only first character of s and return it
+    """
+    if len(s) == 0:
+        return s
+
+    return s[0].upper() + s[1:]
+
+def k8s_api_method(kind, apiversion, method, namespaced=True):
+    """
+    Return bound method that can perform method action on objects of kind in apiversion
+
+    kind should be the same CamelCased string used in `kind` in the Kubernetes object.
+    apiversion should be the same string set in `apiVersion` in the Kubernetes object.
+    method should be 'create', 'delete', 'list' or 'patch'.
+
+    Set namespaced=False for objects that are not in a namespace (like ClusterRoles or
+    Nodes)
+    """
+    if apiversion == 'v1':
+        class_name = 'CoreV1Api'
+    else:
+        parts = apiversion.replace('k8s.io', '').split('/')
+        parts = [''.join([first_upper(p) for p in part.split('.')]) for part in parts]
```

Comment:
General approach makes sense, but this seems a bit brittle.

Reply:
Yeah I agree! This is sortof prototype quality code now, I need to spend a solid day on this soon.